### ISSUE TITLE
Help with my Stickybits not using scroll method correctly in IE11

### ISSUE THREAD

**Issue by adambant** [dollarshaveclub/stickybits#252]
When page loads, sticky item immediately snaps to top of page not waiting for item to scroll to top of page. Also any offsets applied immediately get applied to component not waiting for component to reach top of page.

**Comment by adambant** [comment]
Works fine in all browsers that support position:sticky, pretty much everything except IE browsers.

**Comment by DanielRuf** [comment]
Please provide a codepen so we can test this better.

I guess there is still an issue with the scroll event listener.

**Comment by adambant** [comment]
Ok after retesting, same behavior. Offset applied to component on page load, then upon any scroll action component snaps to top scroll sticky position. Using 3.1.1

Here is a simple codeine to illustrate what I am seeing on my actual page.
https://codepen.io/adambant/pen/EQdBJd

**Comment by yowainwright** [comment]
@adambant thanks for taking the time to illustrate the issue you're seeing! 🙏

Please review, other [issues](https://github.com/dollarshaveclub/stickybits/issues?q=is%3Aissue+is%3Aclosed) where people noted similar problems. Also, check out this [page](https://github.com/dollarshaveclub/stickybits/wiki/Understanding-Stickybits-and-Position-Sticky) in the wiki.

~Thanks

**Comment by adambant** [comment]
Hi yowainwright,

I have tried many combinations of parent element wrappers, letting <body> be the parent element, using immediate wrappers for parent element etc. Can't seem to get the expected behavior desired. I can fix the offset problem by removing the "position: relative" on the secNav div css. However component either immediately sticks to sticky position OR if I have it in a immediate parent wrapper div just scrolls by and doesn't stick at all. Also if I wrap in a parent div this causes the browsers that support "position: sticky" to not work either.

**Comment by yowainwright** [comment]
@adambant I'm having a tough time following the comment above, unfortunately.

In this [codepen I forked](https://codepen.io/yowainwright/pen/Zrmzgp) from the one that was provided, I updated the dom in 1 place and override `position: sticky` to more easily see how `position: fixed` would look. It seems to work as expected.

**Comment by adambant** [comment]
We need secNav menu to be sticky underneath the "someContent" not inside of it. Also I could have used two "fixed" positions if desired. But am looking for secNav menu to be under "someContent" then stick under main nav when scrolled to. But your example just fixes both menus to top of the page.

**Comment by adambant** [comment]
Short version, if you load my codepen in Chrome/Safari/Firefox that is the expected behavior I want when running it in IE11/Edge.

**Comment by yowainwright** [comment]
@adambant please review [`position: sticky`](https://developer.mozilla.org/en-US/docs/Web/CSS/position) on MDN, the [wiki page](https://github.com/dollarshaveclub/stickybits/wiki/Understanding-Stickybits-and-Position-Sticky) that was provided above, and [old Stickybits issues](https://github.com/dollarshaveclub/stickybits/issues?q=is%3Aissue+is%3Aclosed) where similar problems were reported.

**Net-net:** I don't think this is an issue with Stickybits. 😉

**Comment by adambant** [comment]
I would expect Stickybits according to all documentation to properly toggle on and off the "position: fixed" when the item gets to the proper place on the page according to its scroll position. If it doesn't do this, then the plugin isn't any better than just using position:sticky and letting it fail or not work on the browsers that don't support it.

I.e. plugin adds no value to default browser behavior.

**Comment by DanielRuf** [comment]
IE11 does not support stiky so fixed is used there. There might be still issues with the calculation.

**Comment by yowainwright** [comment]
@DanielRuf it is not in this case. It has to do with the dom structure that is being used—which I've tried to explain. 😔

**Comment by DanielRuf** [comment]
I guess in this case it is the wrong parent selector as parameter or similar. At least this example differs from others which introduces other issues. Not yet 100% sure if bug or bad markup.

**Comment by yowainwright** [comment]
@DanielRuf I reached out to you in a Twitter message. Thanks for helping to troubleshoot!

**Comment by adambant** [comment]
`Stickybits.prototype.computeScrollOffsets = function computeScrollOffsets(item) {
  var it = item;
  var p = it.props;
  var parent = it.parent;
  var isCustom = !this.isWin && p.positionVal === 'fixed';
  var isBottom = p.verticalPosition !== 'bottom';
  var scrollElOffset = isCustom ? p.scrollEl.getBoundingClientRect().top : 0;
  var itScrollOffset = parseInt(it.el.getBoundingClientRect().top, 0);

  var stickyStart = isCustom ? parent.getBoundingClientRect().top - scrollElOffset : parent.getBoundingClientRect().top;
  it.offset = itScrollOffset - p.stickyBitStickyOffset;
  it.stickyStart =  it.offset;
  it.stickyStop = isBottom ? stickyStart + parent.offsetHeight - (it.el.offsetHeight + it.offset) : stickyStart + parent.offsetHeight;

  return it;
};`

**Comment by adambant** [comment]
I updated my CodePen with unminimized code and patched/hacked the computerScrollOffsets function with changed calcs. This might not work for 100% of use cases, e.g. where "bottom" is selected etc. But it works well for my use case which is better than it not working at all for my use case. If this helps pin down the specifics that is great.

**Comment by adambant** [comment]
To note in my code change I did not have to add any extra html or parent wrapper classes etc. It works fine with the < body > tag as the parent container. Also I did test using a wrapper container around the entire contents of my page with the "scrollEl" option and it works ok too. Your results might vary.

**Comment by yowainwright** [comment]
@adambant I think you might have a point & I will look 😊!

> ...If Stickybits is off `position: sticky` spec, I will update the calculation! 🎉

This is what I am understanding, btw:

> Calculate the offset of a sticky element within its parent. Do **not** assume that the sticky element is the **first** child or **does not** have an offset.

---

That said, if you re-look at the [Codepen I pasted](https://codepen.io/yowainwright/pen/Zrmzgp) yesterday and update the `html` slighty—like in the Codepen—placing the _Stickybit_ as a direct child of its containing parent, you can have Stickybits working with little effort and can create a complex sticky UI. Even if there needed to be a calculation update in the Stickybits plugin itself, having a Stickybit element be a direct child of it's parent is going to make making a complex Sticky UI much more manageable. 👌

**Comment by adambant** [comment]
But in your CodePen you have the SecNav hard coded in CSS to be a "fixed" position which completely negates the need or use for stickybits. Also we are using a CMS system and putting the secondary nav inside content blocks is not an option for use. So the secNav needs to be outside any other containers except the main page container. 

If you review my CodePen example you see the secNav follows the page scroll until it hits the man nav menu. Then it stops. It is not permanently stuck the to the bottom of the main nav like your CodePen example.

**Comment by yowainwright** [comment]
Please read the comment above `position: fixed` in the Codepen. I work on a mac so adding `position: fixed` within the Codepen allowed me to quickly see what one might see in Internet Explorer 11. 🤓

I have been trying the help @adambant. Some of the comments I've received don't seem to be very open to being helped. Perhaps I'm interpreting the context of the comments incorrectly.

**Comment by adambant** [comment]
I work on a Mac too. :-) But I use VirtualBox to run a VM so I can test on IE11 and other Microsoft browsers. 

https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/

Forcing a component to "position: fixed" on code on a Mac does not simulate an IE11 behavior. Because all browsers support position: fixed, which only serves to NOT let the sticky bits plugin emulate the position: sticky behavior on IE11 or any other browser including Safari/Chrome or FF. 

Yes I think you have been interpreting the context of the comments incorrectly.

I have already fixed my issue by patching the computeScrollOffsets function which I provided the code above. I am continuing in this discussion to help clarify the problem and expected behavior as advertised of this stickybits plugin and to leave enough information for whomever else might be having  similar problems to resolve them as I have. Hence the codepen example, patched function, and other information I have given.
